src/lib/maintenance/db.js

```diff
@@ -0,0 +1,347 @@
+);
+
+/**
+ * Executes a maintenance command on a database table.
+ * @async
+ * @function
+ * @param {string} command - The command to execute on the table.
+ * @param {string} category - The category of the maintenance command.
+ * @param {string} type - The type of the maintenance command.
+ * @param {Object} model - The database model representing the table.
+ * @returns {Promise} A promise that resolves with the result of the maintenance command.
+ */
+const tableMaintenanceCommand = async (
+  command,
+  category,
+  type,
+  model,
+  triggeredById,
+) => {
+  // Get the name of the table from the model
+  const tableName = model.getTableName();
+  // Execute the maintenance command on the table using the maintenanceDBCommand function
+  return maintenanceDBCommand(`${command} "${tableName}";`, category, type, { table: tableName }, triggeredById);
+};
+
+/**
+ * This function performs a vacuum operation on a database table.
+ * @async
+ * @function
+ * @param {string} model - The name of the database table to perform the vacuum operation on.
+ * @returns {Promise<void>}
+ */
+const vacuumTable = async (model, triggeredById = null) => tableMaintenanceCommand(
+  // Execute the tableMaintenanceCommand function with the 'VACUUM FULL' command, maintenance
+  // category DB, maintenance type VACUUM, and the provided model parameter.
+  'VACUUM FULL',
+  MAINTENANCE_CATEGORY.DB,
+  MAINTENANCE_TYPE.VACUUM,
+  model,
+  triggeredById,
+);
+
+/**
+ * Asynchronous function that reindexes a database table.
+ * @param {string} model - The name of the database table to be reindexed.
+ * @returns {Promise} A promise that resolves when the reindexing is complete.
+ */
+const reindexTable = async (model, triggeredById = null) => tableMaintenanceCommand(
+  'REINDEX TABLE', // SQL command to reindex a table
+  MAINTENANCE_CATEGORY.DB, // Maintenance category for database maintenance commands
+  MAINTENANCE_TYPE.REINDEX, // Maintenance type for reindexing
+  model, // Name of the table to be reindexed
+  triggeredById,
+);
+
+/**
+ * This function vacuums all tables in the database using Sequelize ORM.
+ * @param {number} offset - The starting index of the models to vacuum.
+ * @param {number} limit - The number of models to vacuum.
+ * @returns {Promise} - A promise that resolves to an object indicating whether the vacuuming was
+ * successful or not.
+ */
+const vacuumTables = async (offset = 0, limit = numOfModels, triggeredById = null) => {
+  // Get all models from Sequelize and sort them alphabetically by table name.
+  const models = Object.values(sequelize.models)
+    .sort((a, b) => a.getTableName().localeCompare(b.getTableName()))
+    .slice(offset, offset + limit);
+
+  // Call maintenanceCommand function with an async function that vacuums each model's table.
+  return maintenanceCommand(
+    async (logMessages, logBenchmarks, triggered) => ({
+      isSuccessful: (await Promise.all(models.map(async (model) => vacuumTable(
```

Comment:
Given that this runs at night, why don't we just do this sequentially rather than all at once? Then we likely never have to worry about hitting limits. I also don't think we really get anything out of a vacuum full unless we've just run a migration, which we could just put on the tail of a migration that deletes large numbers of records. So, I would suggest we do ordinary VACUUM ANALYZE every night. Tables that have been recently vaccumed are usually skipped unless there's big changes, and if there are big changes, that's probably exactly when we *want* them to be maintained as soon as possible. I would expect that it would take a minute or so for the entire database.

Reply:
I am switching over to VACUUM ANALYZE